### DISCUSSION TITLE
AggregatingMergeTree computed column based on the value of another column

### DISCUSSION THREAD

**Discussion by bash-spbu** [ClickHouse/ClickHouse#66965] [Q&A]
Hello!

I have a question regarding the `AggregatingMergeTree` engine.
Backbone example code is provided here: https://fiddle.clickhouse.com/2daeb5b2-f40b-49ca-9ab0-4bd9ba1a05f7

Speaking briefly, I would formulate it as _how to calculate one column from the results of aggregation of another in `AggregatingMergeTree`_?

Speaking more broadly, consider the following use case:
- We have events from users that contain the time and money amount of a single user's purchase.
- We have an aggregation table that sums purchases by users.
- We want to select users whose total amount of purchases exceeded some threshold during the last 30 minutes (in order to send them a bonus/ discount, for example).

The total amount of users is hundreds of millions, but the amount of users at the last point is likely to be 3-5.

So I was thinking that in order to get results only from the last 30 minutes, I could add a column `accepted_at` that would contain a timestamp when the aggregation amount exceeded the threshold, and then filter by `accepted_at >= (now() - INTERVAL ${ΔT_update})`. Without it, I need to somehow store filter results from previous and current query executions (which result in millions of users) and diff them in order to get only 3-5 resulting rows.

Also, `accepted_at` would be a perfect candidate for minmax index, and CH would read literally only the sporadic needed granules.

But I could not manage to express to CH how and when to calculate `accepted_at`. Materialized columns and views cannot help as they express only the initial column values, and I need it to recalculate on every parts merge. I guess I need UDAF which are not implemented yet...

Could you please give a hint whether such functionality is possible with the current CH or maybe whether there is another recommended way for such tasks?

**Comment by bash-spbu** [comment]
I guess that this proposed feature could help here: https://github.com/ClickHouse/ClickHouse/issues/4419

**Comment by den-crane** [comment]
Do you mean last point is different for different users? 
Like for some user last 30 minutes will in 2015, for another in 2020 ?

Otherwise I just don't understand what is the problem to remove data which is older than 30 minutes. And calculate only over the last 30 minutes.

**Comment by bash-spbu** [comment]
The last point (and whole 30 minutes) are the same for all users. 
But "30 minutes" is only the threshold completion check interval, the whole aggregation window is usually much larger. 

Let me clarify it on the picture below. For example, we want to grant bonuses to all users who made purchases with total amount exceeding some threshold during _the last month_.

![events](https://github.com/user-attachments/assets/628b7007-a02e-4bf0-bfba-96e66097093a)

So on the picture above we have 3 intervals:

| Color | Designation | Time Size | Events Size | Description |
| ------------- | ------------- | ---- | --- | --- |
| red  | `T_all` | 1 year | 10^9 | all recorded events for the whole time |
| yellow | `ΔT_lookback` | 1 month | 10^8 | event window aggegation interval |
| blue | `ΔT_update` | 30 minutes | 10^4 | threshold pass check interval |

Every 30 minutes, we check whether some users have passed the threshold. As the blue arrow represents in the picture, there will be some final event that made a decisive contribution in the last 30 minutes. But prior events (represented by green arrows) could happen throughout the whole month.

So we need to aggregate for events for the whole yellow zone, which is less than the red zone but still orders of magnitude more than the blue zone.

Currently, every 30 minutes, we need to check all tracked users to see if they have passed the threshold. And then to diff them with a set of users who passed the threshold prior to the last 30 minutes. It's quite an amount of work but at least it doesn't depend on the events count, only on the users count, which scales way better.

But if we could add a column that tracked the moment in time when aggregation values passed the threshold, we could easily filter by it with the help of the minmax index, for instance. Then we do not depend not only on the event count, but on the all users count as well.
